### PR TITLE
core: add instrinsics for llvm.expect.i8-i64

### LINKED ISSUE #20579: Implement branch prediction intrinsics
Branch prediction intrinsic equivalent to GCC’s `__builtin_expect` should exist. Alternatively two intrinsics `likely` and `unlikely` could be implemented.

### PR DESCRIPTION
**Pull request by seanmonstar**
Closes #20579

### CODE REVIEW

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -545,3 +565,32 @@ extern "rust-intrinsic" {
     /// Performs checked `u64` multiplication.
     pub fn u64_mul_with_overflow(x: u64, y: u64) -> (u64, bool);
 }
+
+/// Inform the optimizer that `expected_val` is expected
+/// (the most probable) value of val.
+#[cfg(stage0)]
+pub unsafe fn expect8(val: u8, _expected_val: u8) -> u8 {
```
Comment (huonw) [inline src/libcore/intrinsics.rs]: I... expect... these aren't actually necessary to do the bootstrap?

Comment (seanmonstar) [inline src/libcore/intrinsics.rs]: I don't follow? Before i added cfg(stage0) versions, it complained during stage0 that these intrinsics weren't known by the compiler.

Comment (huonw) [inline src/libcore/intrinsics.rs]: These functions aren't called during the bootstrap, so you can have the FFI declarations be `#[cfg(not(stage0))]` and just omit these altogether. (I.e. have `core::intrinsics::expect*` not exist in `stage0` at all.)

Comment (seanmonstar) [inline src/libcore/intrinsics.rs]: Oh I see what you mean.

**Comment by alexcrichton** [comment]
Currently we are trying to hide the `intrinsics` module entirely from the standard library's API so in the long run these will not be accessible. (see https://github.com/rust-lang/rust/pull/22358#issuecomment-74462306 for some more info).

In the past we've been ok adding these form of intrinsics if they're directly used in the standard library and provide tangible benefits. Exposing them as a public API, however, would require an RFC in any case. For now though because these are not used internally and also not exposed through a stable interface, I'm going to close this. I also think that #20579 should live in the RFC repo instead of this repo as it's a general feature request, I'll see if can't get it moved over.

Thanks for the PR though!

**Pull request closed without merging**